---
title: "Differential Gene Expression of *Bacteroides thetaiotaomicron*"
author: "Gina Vazquez"
date: "2023-11-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE, include=FALSE}
library(SummarizedExperiment)
library(tximeta)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(readr)
library(pheatmap)
library(tidyverse)
library(IHW)
library(ashr)
library(ggplot2)
library("RColorBrewer")
library(ggrepel)
```

## Project Overview

The mammalian gut microbiota is influenced by various factors such as alterations in diet, the introduction of drugs, and the onset of diseases. Numerous studies have explored mechanisms through which the gut microbiota responds to these stimuli. For example, in [Zhu et al., 2020](https://pubmed.ncbi.nlm.nih.gov/32075741/) it was found that *Bacteroides thetaiotaomicron* utilizes xenosiderophores to acquire iron within an inflamed gut. Here, I will investigate the differential expression of genes in B. thetaiotamicron in response to a Salmonella Typhimurium (S. Tm) infection in a defined minimal community of native microbiota in murin samples. 

## Project Outline

The differential gene expression of *Bacteroides thetaiotaomicron VPI-5482* in two groups will be compared. Samples 275.1 - 275.5 are biological replicates of a mouse gut (GRCm39) inoculated with *Clostridium symbiosum* and *Bacteroides thetaiotaomicron VPI-5482*. Samples 278.1 - 278.5 are biological replicates of a mouse gut (GRCm39) inoculated with *Clostridium symbiosum, Bacteroides thetaiotaomicron VPI-5482*, and *Salmonella enterica* serovar Typhimurium.
The sequencing data used in this analysis are from two published papers by Sebastian Winter's lab. Note that mouse microbiome samples for [Zhu et al., 2020](https://pubmed.ncbi.nlm.nih.gov/32075741/) and [Spiga et al., 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5599368/) were collected and processed at the same time. These data are not all publicly available through ENA hence why the naming convention of the samples are not tied to corresponding accession numbers. The data were acquired from the lab's box link and I kept the original, in-lab naming convention. When possible, the names of the samples will be matched to the GenBank/ENA accession numbers for public availability.

## Metatranscriptomics Summary

### Snakemake Workflow

(1) Rule fastp

The fastq of the aforementioned ten samples were obtained from a raw data spreadsheet stored in a Box account held by Dr. Sebastian Winter's lab. Fastp was used to trim adapters, bases with a quality phred-score lower than 4, and reads shorter than 31 bp in length. It also performed base corrections during trimming. A fastq file, json file, and html file were produced for each sample.

(2) Rule filter_mouse_reads

Bowtie2 was used to align the trimmed fastq files generated by the fastp rule to the index files of the mouse genome ([GRCm39](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001635.27/)). The index files for the mouse genome were made outside the Snakemake workflow. Reads that did not align to the mouse genome were saved as filtered fastq files.

(3) Rule align_to_microbe

Bowtie2 was used to align the filtered fastq files generated by rule filter_mouse_reads to the index files for the [*Bacteroides thetaiotaomicron* genome](https://www.ncbi.nlm.nih.gov/datasets/taxonomy/818/). The index files for the microbial genome were made outside the Snakemake workflow. Reads aligning to the microbial genome were saved as a sam file.

(4) Rule convert_sam_to_sorted_bam

Samtools was used to convert sam files generated by rule align_to_microbe to sorted bam files.

(5)  FeatureCounts was used to count transcripts that mapped to the microbial genome from the sorted bam files generated by the convert_sam_to_sorted_bam rule. It was not written as a rule in the Snakemake file because of issues with the variables. The FeatureCounts command is noted as a comment in the Snakemake file.

### Future Notes

An alternative approach:

-   Make a "master reference genome" containing all the reference genomes relevant to all samples.
-   Generate index files from the master reference genome.
-   Align all samples to these index files.
-   Generate FeatureCounts table of all reads that mapped.

Deseq2 calculates library size to determine the scaling factor for each sample based on the FeatureCounts table. Creating a FeatureCounts table with all aligned reads would retain the true library size which could effect he differential gene expression results.

**Note** Hannah has made a workflow with these updates. I will update this document with a summary of the workflow and effects on DGE results. 
## Deseq2 Workflow

### Read in FeatureCounts Table

```{r}
# The gene_name column cannot be the row names because there are repeat names 
df_bth_fc = read.delim("bth_1110.txt", skip = 1, header = TRUE, row.names = 1, stringsAsFactors = FALSE)
df_bth = df_bth_fc[c("X275.1_bth.sorted.bam",	"X275.2_bth.sorted.bam",	"X275.3_bth.sorted.bam",	"X275.4_bth.sorted.bam",	"X275.5_bth.sorted.bam",	"X278.1_bth.sorted.bam",	"X278.2_bth.sorted.bam",	"X278.3_bth.sorted.bam",	"X278.4_bth.sorted.bam",	"X278.5_bth.sorted.bam")]

head(df_bth_fc)
```

### Make the DeSeq2 object

The samples and conditions are saved in 'exp_bth.csv'. The FeatureCounts data frame and experimental design CSV are used to make the DeSeq2 object.  
```{r, warning=FALSE}
csv_exp_bth = read.csv("exp_bth.csv")
print(csv_exp_bth)
dds_bth = DESeqDataSetFromMatrix(countData = df_bth, colData = csv_exp_bth, design = ~ condition)
```

## Results

### Differential Gene Expression

```{r, message=FALSE}
# Set factor levels
dds_bth$condition = factor(dds_bth$condition, levels = c("cs/bt","cs/bt/stm"))

# DeSeq2 results 
dds_bth = DESeq(dds_bth)
bth_res = results(dds_bth)

# Match gene names to results 
bth_res_names <- bth_res
bth_res_names$genenames <- df_bth_fc$gene_name[seq_along(bth_res[, 1])]

# Make a table of DeSeq2 results
write.csv(bth_res_names, file= "bth_results.csv")

head(bth_res_names)
summary(bth_res_names)
```

### Sort Results 

```{r}
# Sort results by log2 fold change
res_sorted = bth_res_names[order(abs(bth_res_names$log2FoldChange), decreasing = TRUE), ]

# Subset significant results based on adjusted p-value threshold (e.g., 0.05)
significant_genes_bth = subset(res_sorted, padj < 0.05)

# View summary of significant genes
summary(significant_genes_bth)
head(significant_genes_bth)
write.csv(significant_genes_bth, file = "bth_sorted_results.csv")
```

### MA Plot with Apeglm Shrinkage

More information on [MDS shrinkage](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#moreshrink).
```{r, warning=FALSE, message=FALSE}
# shrink data
apeglm_bth = lfcShrink(dds_bth, coef="condition_cs.bt.stm_vs_cs.bt", type="apeglm")

# Match gene names to results 
ma_bth_names <- as.data.frame(apeglm_bth)
ma_bth_names$genenames <- df_bth_fc$gene_name[seq_along(apeglm_bth[, 1])]

# Save MA Data in table 
write.csv(ma_bth_names, file="bth_ma_results.csv")

# make ma plot
 ggplot(as.data.frame(ma_bth_names), aes(y = log2FoldChange, x = baseMean)) +
  geom_point(size=0.5) +
  labs(y = "Log2 Fold Change (M)", x = "Average Expression (A)") +
  scale_x_continuous(trans='log10') +
  ggtitle("MA Plot of BTH Set with Shrinkage") +
  geom_text_repel(
    data = subset(ma_bth_names, log2FoldChange > 3 | log2FoldChange < -3),
    aes(label=genenames), 
    #aes(label = ifelse(!is.na(genenames), genenames, rownames(ma_bth_names))),
    box.padding = 0.5,
    point.padding = 0.1,
    segment.color = "grey",
    segment.size = 0.2,
    max.overlaps = 5
  )
```

### Normalization with Variance Stabilized Transformation

The `vst()` function performs variance stabilized transformation on the raw count data while controlling for library size which is necessary for an MDS plot.
```{r}
vsd_bth = vst(dds_bth)
```

### Sample Distances

Sample distances are calculated and used for an MDS plot to compare similarity in counts between samples.
```{r}
#sample distances
sample_dists_bth = assay(vsd_bth) %>%
  t() %>%
  dist() %>%
  as.matrix()

#MDS values from distance matrix
mds_data_bth = data.frame(cmdscale(sample_dists_bth))
mds_bth = cbind(mds_data_bth, as.data.frame(colData(vsd_bth)))
```

### PCA Plot

```{r}
# Shorten names of samples 
short_vsd_bth <- vsd_bth
colnames(short_vsd_bth) <- sub("_bth.sorted.bam", "", colnames(vsd_bth))

plotPCA(short_vsd_bth) + 
  geom_text_repel(aes(label = colnames(short_vsd_bth)), size = 3, hjust = 1.3, vjust = -0.5, box.padding = 0.2, segment.size = 0) +
  geom_point(size = 0.1)
```

### Volcano Plot 

```{r, warning=FALSE}
# Make a copy of the deseq2 results data frame to modify for the plot 
bth_res_vol = data.frame(apeglm_bth)

# Add names
bth_res_vol$genenames <- df_bth_fc$gene_name[seq_along(bth_res_vol[, 1])]

# Add a column of NAs
bth_res_vol$diffexpressed <- "NO"

# If log2Foldchange > 0.6 and pvalue < 0.05, set as "UP" 
bth_res_vol$diffexpressed[bth_res_vol$log2FoldChange > 0.6 & bth_res_vol$pvalue < 0.05] <- "UP"

# If log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"
bth_res_vol$diffexpressed[bth_res_vol$log2FoldChange < -0.6 & bth_res_vol$pvalue < 0.05] <- "DOWN"

# Label the volcano plot points
# Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
bth_res_vol$delabel <- NA

# Populate the delabel column with values from the genenames column 
bth_res_vol$delabel[bth_res_vol$diffexpressed != "NO"] <- bth_res_vol$genenames[bth_res_vol$diffexpressed != "NO"]

# Save volcano plot data as a table 
write.csv(bth_res_vol, file="bth_vol_results.csv")

ggplot(data=as.data.frame(bth_res_vol), aes(x = log2FoldChange, y = -log10(pvalue), col=diffexpressed, label=delabel)) +
        geom_point(size=0.5) +
        theme_minimal() +
        geom_text_repel() +
        scale_color_manual(values=c("blue", "black", "red")) +
        geom_vline(xintercept=c(-0.6, 0.6), col="red") +
        geom_hline(yintercept=-log10(0.05), col="red")
```

### HeatMap

```{r}
# transform the raw count data from the DeSeq2 object
bth_ntd = normTransform(dds_bth)

#shorten the sample names in bth_ntd
colnames(bth_ntd) <- sub("_bth.sorted.bam", "", colnames(bth_ntd))
rownames(bth_ntd) <- sub("gene-FE838_", "", rownames(bth_ntd))

# extract the top 30 genes with the highest mean expression 
bth_select = order(rowMeans(counts(dds_bth,normalized=TRUE)),
                   decreasing=TRUE)[1:30]

# Create a subset of normalized and transformed data with only the top 30 genes based on mean expression level
heatmap_bth = (assay(bth_ntd)[bth_select,])

# Match gene names to results 
heatmap_bth_names <- as.data.frame(heatmap_bth)
heatmap_bth_names$genenames <- df_bth_fc$gene_name[seq_along(heatmap_bth[, 1])]

# Assuming heatmap_bth_names is your data frame
rownames_to_use_bth <- rownames(heatmap_bth_names)
row_indices_bth <- grepl("^RS", rownames_to_use_bth)

# Replace row names starting with "RS" if the corresponding genenames value is not NA
rownames_to_use_bth[row_indices_bth] <- ifelse(!is.na(heatmap_bth_names$genenames[row_indices_bth]),
                                       heatmap_bth_names$genenames[row_indices_bth],
                                       rownames_to_use_bth[row_indices_bth])

# Save heatmap data as table
write.csv(heatmap_bth_names, file = "bth_heat_results.csv")

# Use pheatmap with the modified row names
pheatmap(heatmap_bth_names[, !colnames(heatmap_bth_names) %in% c("genenames")],
         cluster_rows = TRUE, show_rownames = TRUE, show_colnames = TRUE, cluster_cols = FALSE,
         labels_row = rownames_to_use_bth)
```

### HeatMap of Sample-to-Sample Distances

```{r}
bth_dist = dist(t(assay(vsd_bth)))

bth_dist_matrix = as.matrix(bth_dist)

rownames(bth_dist_matrix) = paste(vsd_bth$condition, vsd_bth$type)
colnames(bth_dist_matrix) <- sub("_bth.sorted.bam", "", colnames(vsd_bth))

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(bth_dist_matrix,
         clustering_distance_rows=bth_dist,
         clustering_distance_cols=bth_dist,
         col=colors,
         show_colnames = TRUE)
```

## Additional Resources

Further information on how to use DeSeq2 for metatranscriptomic analysis in R can be found [here](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-quality-assessment-by-sample-clustering-and-visualization).
The link to the repository for this project can be found [here](https://github.com/GinaVazquez/Mouse-Metatranscriptomics). Refer to the README.md for list of the different files.

## Appendix 

### MDS Plot

```{r}
# Shorten names of samples 
short_mds_bth <- mds_bth
rownames(short_mds_bth) <- sub("_bth.sorted.bam", "", rownames(mds_bth))

ggplot(short_mds_bth, aes(X1, X2, color = condition)) +
  geom_point(size = 3) +
  geom_text(aes(label = rownames(short_mds_bth)), size = 3, hjust = 1.4, vjust = -0.5) +
  theme_minimal()
```

### MA Plot of BTH group without shrinkage

```{r, warning=FALSE, message=FALSE}
ma_data_bth = data.frame(M = bth_res$log2FoldChange, A = rowMeans(counts(dds_bth, normalized = TRUE)))
ggplot(ma_data_bth, aes(y = M, x = A)) +
  geom_point() +
  labs(y = "Log2 Fold Change (M)", x = "Average Expression (A)") +
  scale_x_continuous(trans='log10') +
  ggtitle("MA Plot of BTH Set")
```