---
title: "Differential Gene Expression of *Bacteroides thetaiotaomicron*"
author: "Gina Vazquez"
date: "2023-11-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE, include=FALSE}
library(SummarizedExperiment)
library(tximeta)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(readr)
library(pheatmap)
library(tidyverse)
library(IHW)
library(ashr)
library(ggplot2)
library("RColorBrewer")
library(ggrepel)
```

## Project Overview

-   paragraph summarizing takeaways from b theta paper and how now we're examining gene expression changes in b theta which cause the phenotypical changes seen in the paper. 

## Project Outline

Here I am comparing the gene expression of *Bacteroides thetaiotaomicron VPI-5482* in two groups. Samples 275.1 - 275.5 are biological replicates of a mouse gut (GRCm39) inoculated with *Clostridium symbiosum* and *Bacteroides thetaiotaomicron VPI-5482*. Samples 278.1 - 278.5 are biological replicates of a mouse gut (GRCm39) inoculated with *Clostridium symbiosum, Bacteroides thetaiotaomicron VPI-5482*, and *Salmonella enterica* serovar Typhimurium.
The sequencing data used in this analysis are from two published papers by Sebastian Winter's lab. Note that mouse microbiome samples for [Zhu et al., 2020](https://pubmed.ncbi.nlm.nih.gov/32075741/) and [Spiga et al., 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5599368/) were collected and processed at the same time. These data are not all publicly available through ENA hence why the naming convention of the samples are not tied to corresponding accession numbers. The data were acquired from the lab's box link and I kept the original, in-lab naming convention. When possible, the names of the samples will be matched to the GenBank/ENA accession numbers for public availability.

## Metatranscriptomics Summary

### Snakemake Workflow

(1) Rule fastp

The fastq of the aforementioned ten samples were obtained from a raw data spreadsheet stored in a Box account held by Dr. Sebastian Winter's lab. Fastp was used to trim adapters, bases with a quality phred-score lower than 4, and reads shorter than 31 bp in length. It also performed base corrections during trimming. A fastq file, json file, and html file were produced for each sample.

(2) Rule filter_mouse_reads

Bowtie2 was used to align the trimmed fastq files generated by the fastp rule to the index files of the mouse genome ([GRCm39](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001635.27/)). The index files for the mouse genome were made outside the Snakemake workflow. Reads that did not align to the mouse genome were saved as filtered fastq files.

(3) Rule align_to_microbe

Bowtie2 was used to align the filtered fastq files generated by rule filter_mouse_reads to the index files for the [*Bacteroides thetaiotaomicron* genome](https://www.ncbi.nlm.nih.gov/datasets/taxonomy/818/). The index files for the microbial genome were made outside the Snakemake workflow. Reads aligning to the microbial genome were saved as a sam file.

(4) Rule convert_sam_to_sorted_bam

Samtools was used to convert sam files generated by rule align_to_microbe to sorted bam files.

(5)  FeatureCounts was used to count transcripts that mapped to the microbial genome from the sorted bam files generated by the convert_sam_to_sorted_bam rule. It was not written as a rule in the Snakemake file because of issues with the variables. The FeatureCounts command is noted as a comment in the Snakemake file.

### Future Notes

An alternative approach:

-   Make a "master reference genome" containing all the reference genomes relevant to all samples.
-   Generate index files from the master reference genome.
-   Align all samples to these index files.
-   Generate FeatureCounts table of all reads that mapped.

Deseq2 calculates library size to determine the scaling factor for each sample based on the FeatureCounts table. Creating a FeatureCounts table with all aligned reads would retain the true library size which could effect he differential gene expression results.

**Note** Hannah has made a workflow with these updates. I will update this document with a summary of the workflow and effects on DGE results. 

## Deseq2 Workflow

### Read in FeatureCounts Table

```{r}
df_bth_fc = read.delim("bth_1110.txt", skip = 1, header = TRUE, row.names = 1, stringsAsFactors = FALSE)
df_bth = df_bth_fc[c("X275.1_bth.sorted.bam",	"X275.2_bth.sorted.bam",	"X275.3_bth.sorted.bam",	"X275.4_bth.sorted.bam",	"X275.5_bth.sorted.bam",	"X278.1_bth.sorted.bam",	"X278.2_bth.sorted.bam",	"X278.3_bth.sorted.bam",	"X278.4_bth.sorted.bam",	"X278.5_bth.sorted.bam")]
```

### Make the DeSeq2 object

The samples and conditions are saved in 'exp_bth.csv'. The FeatureCounts data frame and experimental design CSV are used to make the DeSeq2 object.  
```{r}
csv_exp_bth = read.csv("exp_bth.csv")
dds_bth = DESeqDataSetFromMatrix(countData = df_bth, colData = csv_exp_bth, design = ~ condition)
```

## Results

### Differential Gene Expression

```{r, message=FALSE}
# Set factor levels
dds_bth$condition = factor(dds_bth$condition, levels = c("cs/bt","cs/bt/stm"))

# DeSeq2 results 
dds_bth = DESeq(dds_bth)
bth_res = results(dds_bth)

# Make a table of DeSeq2 results
write.table(bth_res, file= "bth_results.txt")
head(bth_res)
```

### Sort Results 

```{r}
# Sort results by log2 fold change (assuming higher absolute log2 fold change is more interesting)
res_sorted = bth_res[order(abs(bth_res$log2FoldChange), decreasing = TRUE), ]

# Subset significant results based on adjusted p-value threshold (e.g., 0.05)
significant_genes = subset(res_sorted, padj < 0.05)

# View summary of significant genes
summary(significant_genes)
write.table(significant_genes, file = "bth_summary.txt")

# Make a table of gene names
gene_names = row.names(significant_genes)
write.table(gene_names, file = "gene_names.txt", quote = FALSE, row.names = FALSE)
```

### MA Plot of BTH group without shrinkage

```{r, warning=FALSE, message=FALSE}
ma_data_bth = data.frame(M = bth_res$log2FoldChange, A = rowMeans(counts(dds_bth, normalized = TRUE)))
ggplot(ma_data_bth, aes(y = M, x = A)) +
  geom_point() +
  labs(y = "Log2 Fold Change (M)", x = "Average Expression (A)") +
  scale_x_continuous(trans='log10') +
  ggtitle("MA Plot of BTH Set")
```

### MA Plot with Apeglm Shrinkage

More information on [MDS shrinkage](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#moreshrink).
```{r, warning=FALSE, message=FALSE}
apeglm_bth = lfcShrink(dds_bth, coef="condition_cs.bt.stm_vs_cs.bt", type="apeglm")

ggplot(as.data.frame(apeglm_bth), aes(y = log2FoldChange, x = baseMean)) +
  geom_point() +
  labs(y = "Log2 Fold Change (M)", x = "Average Expression (A)") +
  scale_x_continuous(trans='log10') +
  ggtitle("MA Plot of BTH Set with Shrinkage")
```

### Normalization with Variance Stabilized Transformation

The `vst()` function performs variance stabilized transformation on the raw count data while controlling for library size which is necessary for an MDS plot.
```{r}
vsd_bth = vst(dds_bth)
```

### Sample Distances

Sample distances are calculated and used for an MDS plot to compare similarity in counts between samples.
```{r}
#sample distances
sample_dists_bth = assay(vsd_bth) %>%
  t() %>%
  dist() %>%
  as.matrix()

#MDS values from distance matrix
mds_data_bth = data.frame(cmdscale(sample_dists_bth))
mds_bth = cbind(mds_data_bth, as.data.frame(colData(vsd_bth)))
```

### MDS Plot

```{r}
ggplot(mds_bth, aes(X1, X2, color = condition)) + 
  geom_point(size = 3) +
  theme_minimal()
```

### PCA Plot

```{r}
plotPCA(vsd_bth)
```

### Volcano Plot 

```{r, warning=FALSE}
bth_res_vol = data.frame(bth_res)
# add a column of NAs
bth_res_vol$diffexpressed <- "NO"
# if log2Foldchange > 0.6 and pvalue < 0.05, set as "UP" 
bth_res_vol$diffexpressed[bth_res_vol$log2FoldChange > 0.6 & bth_res_vol$pvalue < 0.05] <- "UP"
# if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"
bth_res_vol$diffexpressed[bth_res_vol$log2FoldChange < -0.6 & bth_res_vol$pvalue < 0.05] <- "DOWN"

# Now write down the name of genes beside the points...
# Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
bth_res_vol$delabel <- NA

gene_names <- rownames(bth_res_vol)
new_gene_names <- sub("gene-FE838_", "", gene_names)
bth_res_vol$new_gene_names <- new_gene_names
bth_res_vol$delabel[bth_res_vol$diffexpressed != "NO"] <- bth_res_vol$new_gene_names[bth_res_vol$diffexpressed != "NO"]

ggplot(data=as.data.frame(bth_res_vol), aes(x = log2FoldChange, y = -log10(pvalue), col=diffexpressed, label=delabel)) +
        geom_point() + 
        theme_minimal() +
        geom_text_repel() +
        scale_color_manual(values=c("blue", "black", "red")) +
        geom_vline(xintercept=c(-0.6, 0.6), col="red") +
        geom_hline(yintercept=-log10(0.05), col="red")
```

### HeatMap

```{r}
bth_ntd = normTransform(dds_bth)

# extract the top 20 genes with the highest mean expression 
bth_select = order(rowMeans(counts(dds_bth,normalized=TRUE)),
                   decreasing=TRUE)[1:20]

bth_heat_df = as.data.frame(colData(dds_bth)[,c("condition","sample")])  
pheatmap(assay(bth_ntd)[bth_select,], cluster_rows=TRUE, show_rownames=TRUE, show_colnames=TRUE,
         cluster_cols=FALSE)
```

### HeatMap of Sample-to-Sample Distances

```{r}
bth_dist = dist(t(assay(vsd_bth)))

bth_dist_matrix = as.matrix(bth_dist)
rownames(bth_dist_matrix) = paste(vsd_bth$condition, vsd_bth$type, sep="-")
colnames(bth_dist_matrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(bth_dist_matrix,
         clustering_distance_rows=bth_dist,
         clustering_distance_cols=bth_dist,
         col=colors)
```

### Additional Resources

Further information on how to use DeSeq2 for metatranscriptomic analysis in R can be found [here](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-quality-assessment-by-sample-clustering-and-visualization).
The link to the repository for this project can be found [here](https://github.com/GinaVazquez/Mouse-Metatranscriptomics).